From: David Rosca <nowrep@gmail.com>
Date: Tue, 12 Jul 2016 07:54:04 +0000
Subject: KIconEngine: Fix QIcon::hasThemeIcon always returning true
X-Git-Url: http://quickgit.kde.org/?p=kiconthemes.git&a=commitdiff&h=0abf1b7a148cf6b27caea01a329631e0f1daa983
---
KIconEngine: Fix QIcon::hasThemeIcon always returning true

QIcon::hasThemeIcon(name) checks if QIcon::name() == name,
so icon engine must return empty string when icon doesn't exist.
Also implement IsNullHook for Qt 5.7. Comes with autotest.

REVIEW: 128397
BUG: 365130
---


--- a/src/kiconengine.cpp
+++ b/src/kiconengine.cpp
@@ -116,6 +116,9 @@
 
 QString KIconEngine::iconName() const
 {
+    if (!mIconLoader || !mIconLoader->hasIcon(mIconName)) {
+        return QString();
+    }
     return mIconName;
 }
 
@@ -156,3 +159,15 @@
     return true;
 }
+ 
+void KIconEngine::virtual_hook(int id, void *data)
+{
+#if QT_VERSION >= QT_VERSION_CHECK(5, 7, 0)
+    if (id == QIconEngine::IsNullHook) {
+#else
+    if (id == 3) {
+#endif
+        *reinterpret_cast<bool*>(data) = !mIconLoader || !mIconLoader->hasIcon(mIconName);
+    }
+    QIconEngine::virtual_hook(id, data);
+}

--- a/src/kiconengine.h
+++ b/src/kiconengine.h
@@ -78,6 +78,8 @@
     bool read(QDataStream &in) Q_DECL_OVERRIDE;
     bool write(QDataStream &out) const Q_DECL_OVERRIDE;
 
+    void virtual_hook(int id, void *data) Q_DECL_OVERRIDE;
+
 private:
     //TODO KF6: move those into the d-pointer
     QString mIconName;

